File: Week_02/G20200343030632/Leetcode_104_632.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def maxDepth(self, root: TreeNode) -> int:
        # return self.maxDepth_v1(root)
        return self.maxDepth_v2(root)
    def maxDepth_v1(self, root: TreeNode) -> int:
        return self.helper(root, 0)
    def helper(self, node: TreeNode, level: int):
        if node:
            return max(self.helper(node.right, level + 1), self.helper(node.left, level + 1))
        return level

    def maxDepth_v2(self, root: TreeNode) -> int:
        # 迭代
        queue, res = [root] if root else [], 0
        while queue:
            res += 1
            for i in range(len(queue)):
                cur = queue.pop(0)
                if cur.left:
                    queue.append(cur.left)
                if cur.right:
                    queue.append(cur.right)
        return res

File: Week_02/G20200343030632/Leetcode_104_632.java
import java.util.Deque;
import java.util.LinkedList;


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int maxDepth(TreeNode root) {
        // return maxDepth_v1(root);
        return maxDepth_v2(root);
    }

    public int maxDepth_v1(TreeNode root) {
        if (root == null) {
            return 0;
        }
        return Math.max(maxDepth_v1(root.left), maxDepth_v1(root.right)) + 1;
    }
    public int maxDepth_v2(TreeNode root) {
        // 迭代
        int res = 0;
        if (root == null) {
            return res;
        }
        Deque<TreeNode> queue = new LinkedList<>();
        queue.offerLast(root);
        while (!queue.isEmpty()) {
            res ++;
            int sz = queue.size();
            for (int i = 0; i < sz; i++) {
                TreeNode cur = queue.pollFirst();
                if (cur.left != null) {
                    queue.offerLast(cur.left);
                }
                if (cur.right != null) {
                    queue.offerLast(cur.right);
                }
            }
        }
        return res;
    }
}